src/Microsoft.AspNet.Mvc.Core/ControllerActionInvoker.cs

```diff
@@ -59,6 +59,9 @@ protected override object CreateInstance()
 
         protected override void ReleaseInstance(object instance)
         {
+            var serviceProvider = ActionContext.HttpContext.RequestServices;
+            var tempData = serviceProvider.GetRequiredService<ITempDataDictionary>();
+            tempData.Save();
```

Comment:
CAI is already created per-action, so there's no need for a GetService call here, you can thread this through.

Also what happened to a separate provider and dictionary? I expected this code to look like:

```
if (_tempDataProvider.IsTempDataCreated)
{
    _tempDataProvider.Save();
}
```